In [1]:
import pandas as pd

In [8]:
transfer_df = pd.read_csv('data/Top_8_leagues_past_6_windows.csv', index=False)

In [11]:
transfer_df.drop(labels="Unnamed: 0", axis=1, inplace=True)

In [13]:
transfer_df.head()

player  age nationality            position  \
0   Ante Palaversa   18     Croatia  Defensive Midfield   
1       Ko Itakura   21       Japan         Centre-Back   
2   Yangel Herrera   20   Venezuela    Central Midfield   
3    Marlos Moreno   22    Colombia        Right Winger   
4  Anthony Cáceres   26   Australia    Central Midfield   

                   selling_club previous_league est_market_value  \
0              HNK Hajduk Split         Croatia          £495Th.   
1             Kawasaki Frontale           Japan          £630Th.   
2              New York City FC   United States          £900Th.   
3  Clube de Regatas do Flamengo          Brazil          £450Th.   
4             Melbourne City FC       Australia          £450Th.   

                       fee      buying_club window  year buying_league  
0                   £5.67m  Manchester City  s_w=w  2018           GB1  
1                  £990Th.  Manchester City  s_w=w  2018           GB1  
2  End of loanDec 31, 2018  Manchester City  s_w=w  2018           GB1  
3  End of loanDec 31, 2018  Manchester City  s_w=w  2018           GB1  
4  End of loanDec 31, 2018  Manchester City  s_w=w  2018           GB1

In [14]:
transfer_df['fee'].value_counts()

loan transfer              1180
free transfer              1178
-                          1043
End of loanJun 30, 2019     988
End of loanJun 30, 2018     932
                           ... 
£4.12m                        1
£17Th.                        1
End of loanSep 11, 2020       1
End of loanAug 14, 2020       1
End of loanJul 21, 2019       1
Name: fee, Length: 627, dtype: int64

### Counting and labeling loan deals

In [32]:
transfer_df['loan'] = transfer_df['fee'].apply(lambda x: 1 if 'loan' in x.lower() else 0)

In [33]:
transfer_df['loan'].value_counts()

1    5186
0    4480
Name: loan, dtype: int64

5186 loan deals.  Not unexpected.

### Counting and labeling free transfers 

In [22]:
transfer_df['free'] = transfer_df['fee'].apply(lambda x: 1 if 'free' in x.lower() else 0)

In [23]:
transfer_df['free'].value_counts()

0    8488
1    1178
Name: free, dtype: int64

In [28]:
#Want to see if there are any other types of free transfers
# pd.options.display.max_rows = 1200
transfer_df.loc[transfer_df['free'] == 1, 'fee']

20      free transfer
33      free transfer
46      free transfer
86      free transfer
131     free transfer
            ...      
9638    free transfer
9652    free transfer
9653    free transfer
9654    free transfer
9655    free transfer
Name: fee, Length: 1178, dtype: object

Free transfers happen when a contract is run down; contract length is an important determinant of transfer fee but I don't have that information.  Here, while it would be important for clubs to know how much time is left on a contract I'll use the model to bench mark the going market rate for a player of that profile.

### In the fee value count: 1043 "-"
### These appear to be internal promotions, e.g., from the U23 team. Investigating and labeling here

In [38]:
pd.options.display.max_rows = 1200
transfer_df.loc[transfer_df['fee'] == "-", ['player','fee','selling_club','buying_club']]

player fee                  selling_club  \
16            Callum Hudson-Odoi   -                Chelsea FC U23   
19                 Eddie Nketiah   -                Arsenal FC U23   
29                Sean Longstaff   -          Newcastle United U23   
42                   Kyle Taylor   -           AFC Bournemouth U21   
44                   Samir Nasri   -              Disqualification   
63                    Yan Valery   -            Southampton FC U23   
64               Michael Obafemi   -            Southampton FC U23   
93                Arijanet Muric   -           Manchester City U23   
94                   Manu García   -           Manchester City U23   
112                 Oliver Skipp   -         Tottenham Hotspur U23   
118                    Ryan Kent   -              Liverpool FC U23   
119                 Harry Wilson   -              Liverpool FC U23   
125               Rhian Brewster   -              Liverpool FC U23   
156                  Dejan Iliev   -                Arsenal FC U23   
166                Dwight McNeil   -                Burnley FC U23   
225                 Mark Travers   -           AFC Bournemouth U21   
226                 Sam Surridge   -           AFC Bournemouth U21   
242               Grady Diangana   -           West Ham United U23   
286                    Ben White   -    Brighton & Hove Albion U23   
362                Neco Williams   -              Liverpool FC U23   
369                Billy Gilmour   -                Chelsea FC U23   
374              Japhet Tanganga   -         Tottenham Hotspur U23   
375               Alfie Whiteman   -         Tottenham Hotspur U23   
376                  Michel Vorm   -                  Without Club   
390                   Oskar Buur   -   Wolverhampton Wanderers U23   
395               Anthony Gordon   -                Everton FC U23   
420               Will Smallbone   -              Southampton FC B   
445                 Jack Rodwell   -                  Without Club   
465                  Eric García   -           Manchester City U23   
480            Caoimhin Kelleher   -              Liverpool FC U23   
481           Nathaniel Phillips   -              Liverpool FC U23   
488                  Mason Mount   -                Chelsea FC U23   
489                  Reece James   -                Chelsea FC U23   
490                Jamie Cumming   -                Chelsea FC U23   
505                  Michel Vorm   -                  Without Club   
513                 Reiss Nelson   -                Arsenal FC U23   
514                  Joe Willock   -                Arsenal FC U23   
515                  Bukayo Saka   -                Arsenal FC U23   
524             Brandon Williams   -         Manchester United U18   
525              Mason Greenwood   -         Manchester United U18   
526                 Tahith Chong   -         Manchester United U23   
614            Matthew Longstaff   -          Newcastle United U23   
623               Nnamdi Ofoborh   -           AFC Bournemouth U21   
646               Aaron Connolly   -    Brighton & Hove Albion U23   
647                Steven Alzate   -    Brighton & Hove Albion U23   
667                    Adam Idah   -              Norwich City U23   
697                Matija Sarkic   -               Aston Villa U23   
706                Danny Welbeck   -                  Without Club   
708               Richard Nartey   -                  Without Club   
751             Emile Smith Rowe   -                Arsenal FC U23   
769               Jayson Molumby   -    Brighton & Hove Albion U23   
771               Robert Sánchez   -    Brighton & Hove Albion U23   
776                Danny Welbeck   -                  Without Club   
790                  Jimmy Dunne   -                Burnley FC U23   
800           Jake Clarke-Salter   -                Chelsea FC U23   
803                Nathan Baxter   -                Chelsea FC U23   
826              Tyrick Mitchell   -

The fee "-" are internal promotions or picking up players without a club for free.  Will not be used in our model

In [39]:
transfers_reduced = transfer_df.loc[(transfer_df['fee'] != "-")&(transfer_df['loan'] == 0)&(transfer_df['free'] == 0), :]

In [40]:
transfers_reduced.head()

player  age    nationality            position  \
0      Ante Palaversa   18        Croatia  Defensive Midfield   
1          Ko Itakura   21          Japan         Centre-Back   
10  Christian Pulisic   20  United States         Left Winger   
27     Miguel Almirón   24       Paraguay  Attacking Midfield   
38    Dominic Solanke   21        England      Centre-Forward   

         selling_club previous_league est_market_value      fee  \
0    HNK Hajduk Split         Croatia          £495Th.   £5.67m   
1   Kawasaki Frontale           Japan          £630Th.  £990Th.   
10  Borussia Dortmund         Germany          £45.00m  £57.60m   
27  Atlanta United FC   United States          £13.50m  £21.60m   
38       Liverpool FC         England           £9.00m  £19.08m   

         buying_club window  year buying_league  loan  free  
0    Manchester City  s_w=w  2018           GB1     0     0  
1    Manchester City  s_w=w  2018           GB1     0     0  
10        Chelsea FC  s_w=w  2018           GB1     0     0  
27  Newcastle United  s_w=w  2018           GB1     0     0  
38   AFC Bournemouth  s_w=w  2018           GB1     0     0

In [41]:
transfers_reduced.shape

(2259, 14)

In [47]:
# transfers_reduced['previous_league'].value_counts()
# Italy                   307
# France                  267
# England                 232
# Spain                   230
# Portugal                183
# Germany                 177

Evaluating the value counts of previous league - there are only 1213 observations where a player was bought from a big 5 club (1046 where they were not).  Currently only have details statistical profile of players from the top 5 leagues.  Looked to pull in more years and data from more leagues but fbref doesn't have detailed stats for leagues outside the top 5 (except for MLS) or for more than 3 years.  Will have to proceed with what I have understanding this is a bit reduced in scope by virture of data.

In [59]:
transfers_reduced.loc[transfers_reduced['previous_league'] == 'Italy','buying_top_5'] = 1
transfers_reduced.loc[transfers_reduced['previous_league'] == 'France','buying_top_5'] = 1
transfers_reduced.loc[transfers_reduced['previous_league'] == 'England','buying_top_5'] = 1
transfers_reduced.loc[transfers_reduced['previous_league'] == 'Spain','buying_top_5'] = 1
transfers_reduced.loc[transfers_reduced['previous_league'] == 'Germany','buying_top_5'] = 1

/Users/ericroberts/anaconda3/envs/learn-env/lib/python3.6/site-packages/pandas/core/indexing.py:494: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s


In [57]:
transfers_reduced['buying_top_5'].value_counts()

1    1213
0    1046
Name: buying_top_5, dtype: int64

In [78]:
transfers_reduced2 = transfers_reduced.loc[(transfers_reduced['buying_top_5'] == 1), :]

In [79]:
transfers_reduced2.shape

(1213, 15)

### Turning fee column into an integer

In [80]:
transfers_reduced2['fee'].value_counts()

?           75
£1.80m      67
£2.70m      46
£900Th.     42
£3.60m      41
£4.50m      41
£1.35m      38
£7.20m      32
£450Th.     30
£6.30m      30
£9.00m      28
£5.40m      28
£2.25m      26
£8.10m      24
£13.50m     22
£18.00m     21
£10.80m     19
£4.05m      17
£22.50m     15
£3.15m      14
£225Th.     13
£990Th.     13
£16.20m     12
£27.00m     11
£9.90m      11
£270Th.     10
£5.85m      10
£31.50m     10
£4.95m       9
£1.17m       9
£360Th.      8
£11.70m      8
£19.80m      8
£6.75m       8
£1.53m       7
£14.40m      7
£21.60m      7
£14.85m      6
£23.40m      6
£90Th.       6
£12.60m      6
£1.08m       6
£18.90m      6
£12.15m      6
£9.63m       5
£2.97m       5
£25.20m      5
£36.00m      5
£20.70m      5
£720Th.      5
£2.52m       5
£180Th.      5
£40.50m      4
£45.00m      4
£2.43m       4
£72.00m      4
£810Th.      4
£1.62m       4
£1.49m       3
£24.75m      3
£3.96m       3
£11.25m      3
£13.05m      3
£7.65m       3
£6.03m       3
£2.07m       3
£630Th.   

In [81]:
filt =  transfers_reduced2['fee']=='?'
transfers_reduced2.drop(index=transfers_reduced2[filt].index, inplace=True)

/Users/ericroberts/anaconda3/envs/learn-env/lib/python3.6/site-packages/pandas/core/frame.py:4117: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [84]:
transfers_reduced2['fee'].value_counts()

£1.80m      67
£2.70m      46
£900Th.     42
£3.60m      41
£4.50m      41
£1.35m      38
£7.20m      32
£450Th.     30
£6.30m      30
£9.00m      28
£5.40m      28
£2.25m      26
£8.10m      24
£13.50m     22
£18.00m     21
£10.80m     19
£4.05m      17
£22.50m     15
£3.15m      14
£225Th.     13
£990Th.     13
£16.20m     12
£27.00m     11
£9.90m      11
£270Th.     10
£31.50m     10
£5.85m      10
£4.95m       9
£1.17m       9
£360Th.      8
£11.70m      8
£19.80m      8
£6.75m       8
£21.60m      7
£1.53m       7
£14.40m      7
£1.08m       6
£12.15m      6
£14.85m      6
£12.60m      6
£90Th.       6
£23.40m      6
£18.90m      6
£36.00m      5
£25.20m      5
£2.97m       5
£720Th.      5
£180Th.      5
£9.63m       5
£2.52m       5
£20.70m      5
£72.00m      4
£2.43m       4
£40.50m      4
£810Th.      4
£1.62m       4
£45.00m      4
£2.07m       3
£7.65m       3
£11.25m      3
£630Th.      3
£3.96m       3
£6.03m       3
£54.00m      3
£24.75m      3
£1.49m       3
£13.05m   

In [85]:
transfers_reduced2['currency'] = transfers_reduced2['fee'].apply(lambda x: x[0])

/Users/ericroberts/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [87]:
transfers_reduced2['currency'].value_counts()

£    1138
Name: currency, dtype: int64

In [90]:
transfers_reduced2['multiplier'] = transfers_reduced2['fee'].str.extract(r'([a-zA-Z]+)')
                                                                         

/Users/ericroberts/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [100]:
transfers_reduced2['multiplier'].value_counts()

m     964
Th    171
Name: multiplier, dtype: int64

In [102]:
transfers_reduced2.isna().sum()

player              0
age                 0
nationality         0
position            0
selling_club        0
previous_league     0
est_market_value    0
fee                 0
buying_club         0
window              0
year                0
buying_league       0
loan                0
free                0
buying_top_5        0
currency            0
multiplier          3
fee_numerical       0
dtype: int64

In [93]:
transfers_reduced2['fee_numerical'] = transfers_reduced2['fee'].str.extract(r'([0-9]+\.?[0-9]+)')


/Users/ericroberts/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [98]:
transfers_reduced2['fee_numerical'].astype(float)

10       57.60
38       19.08
39       12.24
57        1.53
65       18.90
78       15.30
79        3.11
89       61.02
104       1.53
114      56.25
115      54.00
116      40.50
117      13.23
129      72.00
130      51.30
147      25.79
148      22.50
149      14.40
150       7.20
162      15.21
163      10.98
164       3.51
167      35.28
168      27.23
169      18.18
184      22.50
186      18.99
188      12.60
189      12.60
190       3.60
197       9.63
198       6.30
199       6.03
201       3.60
214       9.63
218      25.20
219      13.50
220      10.17
230      34.20
231      22.50
232      18.00
233       7.20
234       3.96
247      11.70
248       4.50
249       2.52
251       1.80
252       1.53
253     990.00
269      15.21
270       9.00
271       6.30
272       5.40
273       4.05
274       3.20
290      18.00
291       9.00
292       5.85
293       4.50
294       2.25
302      22.50
304      10.17
313      18.00
315      12.60
317      10.26
318       5.04
321       

In [103]:
transfers_reduced2.loc[transfers_reduced2['multiplier'] == 'm', 'fee_final'] = transfers_reduced2['fee_numerical']*1000000
transfers_reduced2.loc[transfers_reduced2['multiplier'] == 'Th', 'fee_final'] = transfers_reduced2['fee_numerical']*1000

/Users/ericroberts/anaconda3/envs/learn-env/lib/python3.6/site-packages/pandas/core/indexing.py:376: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[key] = _infer_fill_value(value)
/Users/ericroberts/anaconda3/envs/learn-env/lib/python3.6/site-packages/pandas/core/indexing.py:494: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s


In [104]:
transfers_reduced2.head()

player  age    nationality        position       selling_club  \
10  Christian Pulisic   20  United States     Left Winger  Borussia Dortmund   
38    Dominic Solanke   21        England  Centre-Forward       Liverpool FC   
39       Chris Mepham   21          Wales     Centre-Back       Brentford FC   
57       Karlan Grant   21        England  Centre-Forward  Charlton Athletic   
65         Jonny Otto   24          Spain       Left-Back    Atlético Madrid   

   previous_league est_market_value      fee              buying_club window  \
10         Germany          £45.00m  £57.60m               Chelsea FC  s_w=w   
38         England           £9.00m  £19.08m          AFC Bournemouth  s_w=w   
39         England           £7.20m  £12.24m          AFC Bournemouth  s_w=w   
57         England          £180Th.   £1.53m        Huddersfield Town  s_w=w   
65           Spain          £10.80m  £18.90m  Wolverhampton Wanderers  s_w=w   

    year buying_league  loan  free  buying_top_5 currency multiplier  \
10  2018           GB1     0     0             1        £          m   
38  2018           GB1     0     0             1        £          m   
39  2018           GB1     0     0             1        £          m   
57  2018           GB1     0     0             1        £          m   
65  2018           GB1     0     0             1        £          m   

   fee_numerical                                          fee_final  
10         57.60  57.6057.6057.6057.6057.6057.6057.6057.6057.605...  
38         19.08  19.0819.0819.0819.0819.0819.0819.0819.0819.081...  
39         12.24  12.2412.2412.2412.2412.2412.2412.2412.2412.241...  
57          1.53  1.531.531.531.531.531.531.531.531.531.531.531....  
65         18.90  18.9018.9018.9018.9018.9018.9018.9018.9018.901...

In [106]:
transfers_reduced2['fee_final']

KeyboardInterrupt: 